# Módulo 3

In [4]:
# importando bibliotecas
import pandas as pd
!pip install openpyxl

In [64]:
# Importando os dados
dados = pd.read_excel('planilha_modulo3.xlsx', engine="openpyxl")

## Analisando brevemente a importação dos dados

In [6]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4271 entries, 0 to 4270
Data columns (total 28 columns):
 #   Column                                                                                                                     Non-Null Count  Dtype  
---  ------                                                                                                                     --------------  -----  
 0   Unnamed: 0                                                                                                                 4271 non-null   int64  
 1   ID                                                                                                                         4271 non-null   object 
 2   IDADE                                                                                                                      4197 non-null   float64
 3   FAIXA IDADE                                                                                                                4271 

In [7]:
dados.describe()

Unnamed: 0        IDADE  MUDOU DE ESTADO?      GESTOR?        SALARIO
count  4271.000000  4197.000000       4171.000000  3694.000000    3694.000000
mean   2135.000000    31.169168          0.806281     0.193016   10969.266919
std    1233.075829     6.902803          0.395258     0.394719   19419.753189
min       0.000000    18.000000          0.000000     0.000000      35.000000
25%    1067.500000    26.000000          1.000000     0.000000    4269.500000
50%    2135.000000    30.000000          1.000000     0.000000    7625.500000
75%    3202.500000    35.000000          1.000000     0.000000   12778.500000
max    4270.000000    54.000000          1.000000     1.000000  485191.000000

## Análise dos dados

Análise de gêneros:

In [8]:
#opções disponíveis na coluna GENERO
dados['GENERO'].unique()

array(['Masculino', 'Feminino', nan, 'Prefiro não informar'], dtype=object)

In [9]:
qtd_fem = len(dados[dados['GENERO']=='Feminino'])
qtd_mas = len(dados[dados['GENERO']=='Masculino'])
qtd_pni = len(dados[dados['GENERO']=='Prefiro não informar'])
qtd_tot = len(dados)


In [10]:
print('Em um total de ', qtd_tot, 'respostas, temos', qtd_fem, 'que assinalaram Feminino', qtd_mas, 'que assinalaram Masculino,', qtd_pni, 'que preferiram não informar e', qtd_tot - qtd_fem - qtd_mas - qtd_pni, 'que deixaram o campo em branco.')

Em um total de  4271 respostas, temos 1056 que assinalaram Feminino 3194 que assinalaram Masculino, 12 que preferiram não informar e 9 que deixaram o campo em branco.


## Atividade
📌 Que tal fazer um filtro para identificar pessoas que se declararam amarelas e outro de pessoas abaixo de 40 anos? E melhor, um filtro que misture ambos?😉   

In [ ]:
# filtro das pessoas amarelas
amarelas = dados['COR/RACA/ETNIA']=='Amarela'

In [28]:
# dados apenas das pessoas amarelas
dados[amarelas].head(3)

Unnamed: 0                                ID  IDADE FAIXA IDADE  \
13          13  zwh83jpy6x40b95xigsr2zwh83jpflu3   24.0       22-24   
85          85  zfeuk2m9jozahicujlla8zfeuk2g7apj   29.0       25-29   
97          97  zbukfwxvpwlk6loglm0zbukfwx6bvyb6   40.0       40-44   

       GENERO COR/RACA/ETNIA  PCD  \
13  Masculino        Amarela  Não   
85   Feminino        Amarela  Não   
97  Masculino        Amarela  Não   

                 EXPERIENCIA_PROFISSIONAL_PREJUDICADA ASPECTOS_PREJUDICADOS  \
13  Não acredito que minha experiência profissiona...                   NaN   
85  Não acredito que minha experiência profissiona...                   NaN   
97  Não acredito que minha experiência profissiona...                   NaN   

    VIVE_NO_BRASIL  ...               SETOR NUMERO DE FUNCIONARIOS GESTOR?  \
13            True  ...  Finanças ou Bancos         Acima de 3.000     0.0   
85            True  ...                 NaN                    NaN     NaN   
97            True  ...              Varejo         Acima de 3.000     0.0   

    CARGO COMO GESTOR                            CARGO ATUAL   NIVEL  \
13                NaN  Analista de Negócios/Business Analyst  Júnior   
85                NaN                                    NaN     NaN   
97                NaN              Analista de BI/BI Analyst  Sênior   

                     FAIXA SALARIAL  \
13   de R$ 4.001/mês a R$ 6.000/mês   
85                              NaN   
97  de R$ 8.001/mês a R$ 12.000/mês   

   QUANTO TEMPO DE EXPERIÊNCIA NA ÁREA DE DADOS VOCÊ TEM?  \
13                                      de 1 a 2 anos       
85                                                NaN       
97                                      de 4 a 6 anos       

   QUANTO TEMPO DE EXPERIÊNCIA NA ÁREA DE TI/ENGENHARIA DE SOFTWARE VOCÊ TEVE ANTES DE COMEÇAR A TRABALHAR NA ÁREA DE DADOS?  \
13                                     Menos de 1 ano                                                                          
85                                                NaN                                                                          
97                                    Mais de 10 anos                                                                          

    SALARIO  
13   5296.0  
85      NaN  
97  11388.0  

[3 rows x 28 columns]

In [30]:
# filtro das pessoas menores ou igual a 40 anos
le40 = dados['IDADE']<= 40

In [31]:
# dados apenas das pessoas com 40 anos ou menos
dados[le40].head(3)

Unnamed: 0                                ID  IDADE FAIXA IDADE     GENERO  \
0           0  zzqzz3l9ily8nuo2m7wyzzqzz3w48o96   39.0       35-39  Masculino   
1           1  zzls2oftfn9law393oezzls2ofhvfpzd   32.0       30-34  Masculino   
3           3  zzbqh3uy7yk7k9qmkzzbqtb4s9faqspl   27.0       25-29  Masculino   

  COR/RACA/ETNIA  PCD               EXPERIENCIA_PROFISSIONAL_PREJUDICADA  \
0          Parda  Não  Não acredito que minha experiência profissiona...   
1          Parda  Não  Sim, acredito que a minha a experiência profis...   
3         Branca  Não                                                NaN   

                          ASPECTOS_PREJUDICADOS  VIVE_NO_BRASIL  ...  \
0                                           NaN            True  ...   
1  Aprovação em processos seletivos/entrevistas            True  ...   
3                                           NaN            True  ...   

              SETOR NUMERO DE FUNCIONARIOS GESTOR?       CARGO COMO GESTOR  \
0     Setor Público         Acima de 3.000     0.0                     NaN   
1       Outra Opção         Acima de 3.000     1.0  Supervisor/Coordenador   
3  Setor Automotivo         Acima de 3.000     0.0                     NaN   

                           CARGO ATUAL   NIVEL  \
0  DBA/Administrador de Banco de Dados  Júnior   
1                                  NaN     NaN   
3    Cientista de Dados/Data Scientist  Sênior   

                     FAIXA SALARIAL  \
0   de R$ 8.001/mês a R$ 12.000/mês   
1    de R$ 4.001/mês a R$ 6.000/mês   
3  de R$ 12.001/mês a R$ 16.000/mês   

  QUANTO TEMPO DE EXPERIÊNCIA NA ÁREA DE DADOS VOCÊ TEM?  \
0                                      de 1 a 2 anos       
1                                      de 3 a 4 anos       
3                                      de 4 a 6 anos       

  QUANTO TEMPO DE EXPERIÊNCIA NA ÁREA DE TI/ENGENHARIA DE SOFTWARE VOCÊ TEVE ANTES DE COMEÇAR A TRABALHAR NA ÁREA DE DADOS?  \
0                                     de 7 a 10 anos                                                                          
1                                     Menos de 1 ano                                                                          
3  Não tive experiência na área de TI/Engenharia ...                                                                          

   SALARIO  
0  11194.0  
1   4695.0  
3  15463.0  

[3 rows x 28 columns]

In [35]:
# dados com os filtros combinados
dados[amarelas & le40]

Unnamed: 0                                ID  IDADE FAIXA IDADE  \
13            13  zwh83jpy6x40b95xigsr2zwh83jpflu3   24.0       22-24   
85            85  zfeuk2m9jozahicujlla8zfeuk2g7apj   29.0       25-29   
97            97  zbukfwxvpwlk6loglm0zbukfwx6bvyb6   40.0       40-44   
159          159  ytrnz3zyvn791t3dum1ytrnz3zp22tmk   37.0       35-39   
170          170  yri3z491ejizlpj9ylu2k3jyri3z4915   34.0       30-34   
...          ...                               ...    ...         ...   
3947        3947  2tnsh0gxdu3aq4z2d2tnsh0hk15brx7q   31.0       30-34   
4058        4058  1t3ndwx8j50k0bz1t3ndwxdpwyg17c2k   27.0       25-29   
4176        4176  0qpa64gqpieg60xpygnajt70qpa64g1a   27.0       25-29   
4207        4207  0ibb8fj9qrwpswfqiil0ibb89e3e4hz3   30.0       30-34   
4262        4262  02vlm03c29g3fcys8tj06102vlm0983h   31.0       30-34   

         GENERO COR/RACA/ETNIA  PCD  \
13    Masculino        Amarela  Não   
85     Feminino        Amarela  Não   
97    Masculino        Amarela  Não   
159    Feminino        Amarela  Não   
170   Masculino        Amarela  Não   
...         ...            ...  ...   
3947   Feminino        Amarela  Não   
4058   Feminino        Amarela  Não   
4176   Feminino        Amarela  Não   
4207   Feminino        Amarela  Não   
4262  Masculino        Amarela  Não   

                   EXPERIENCIA_PROFISSIONAL_PREJUDICADA  \
13    Não acredito que minha experiência profissiona...   
85    Não acredito que minha experiência profissiona...   
97    Não acredito que minha experiência profissiona...   
159   Sim, acredito que a minha a experiência profis...   
170   Não acredito que minha experiência profissiona...   
...                                                 ...   
3947  Sim, acredito que a minha a experiência profis...   
4058  Sim, acredito que a minha a experiência profis...   
4176  Não acredito que minha experiência profissiona...   
4207  Sim, acredito que a minha a experiência profis...   
4262  Não acredito que minha experiência profissiona...   

                                  ASPECTOS_PREJUDICADOS  VIVE_NO_BRASIL  ...  \
13                                                  NaN            True  ...   
85                                                  NaN            True  ...   
97                                                  NaN            True  ...   
159   Oportunidades de progressão de carreira, Aprov...            True  ...   
170                                                 NaN            True  ...   
...                                                 ...             ...  ...   
3947  Quantidade de oportunidades de emprego/vagas r...           False  ...   
4058  Nível de cobrança no trabalho / Stress no trab...            True  ...   
4176                                                NaN            True  ...   
4207  Oportunidades de progressão de carreira, Veloc...            True  ...   
4262                                                NaN            True  ...   

                               SETOR NUMERO DE FUNCIONARIOS GESTOR?  \
13                Finanças ou Bancos         Acima de 3.000     0.0   
85                               NaN                    NaN     NaN   
97                            Varejo         Acima de 3.000     0.0   
159   Tecnologia/Fábrica de Software           de 101 a 500     0.0   
170                           Varejo         Acima de 3.000     1.0   
...                              ...                    ...     ...   
3947  Tecnologia/Fábrica de Software         Acima de 3.000     1.0   
4058  Tecnologia/Fábrica de Software         Acima de 3.000     0.0   
4176                             NaN                    NaN     NaN   
4207             Área de Consultoria         de 501 a 1.000     0.0   
4262  Tecnologia/Fábrica de Software         de 501 a 1.000     1.0   

            CARGO COMO GESTOR  \
13                        NaN   
85                        NaN   
97                        NaN   
159  

------------------


## Exercício de Revisão | Funções

Baixe o arquivo licitacoes.csv e verifique quantas licitações estão com a “Situação Licitação” igual a “Encerrado”. 

In [45]:
# detectar a codificação
!pip install chardet
import chardet

# Lê uma amostra do arquivo (em modo binário)
with open('licitacoes.csv', 'rb') as f:
    resultado = chardet.detect(f.read(10000))  # Lê os primeiros 10 mil bytes

print(resultado)

{'encoding': 'ISO-8859-1', 'confidence': 0.73, 'language': ''}


In [50]:
licitacoes = pd.read_csv('licitacoes.csv', encoding='ISO-8859-1', sep=';')

In [55]:
qtd_encerrado = len(licitacoes[licitacoes['Situação Licitação']=='Encerrado'])

In [58]:
print('Tem ',qtd_encerrado,' licitações encerradas', sep='')

Tem 9016 licitações encerradas


In [61]:
licitacoes['Situação Licitação'].value_counts()

Situação Licitação
Encerrado                         9016
Publicado                         3709
Evento de Resultado de Julgame    1922
Retificação                        127
Evento de Alteração Publicad        93
Evento de Suspensão Publicado       72
Anulação                            67
Revogação                           63
Divulgado                           55
Evento de Adiamento Publicado       46
Evento de Retificação Public        45
Evento de Revogação Publicad        40
Evento de Alteração de Resul        37
Evento de Anulação Publicado        18
Evento de Retificação Divulg         7
Inválido                             4
Evento de Habilitação Public         4
Evento de Alteração Divulgad         4
Evento de Reabertura de Prazo        4
Evento de Adiamento Divulgado        3
Name: count, dtype: int64

## Agrupamento

### GroupBy

In [66]:
dados.groupby('GENERO')['ID'].nunique()

GENERO
Feminino                1055
Masculino               3194
Prefiro não informar      12
Name: ID, dtype: int64

In [69]:
dados.groupby('GENERO', dropna=False)['ID'].nunique()

GENERO
Feminino                1055
Masculino               3194
Prefiro não informar      12
NaN                        9
Name: ID, dtype: int64

### Value Counts

In [70]:
dados['GENERO'].value_counts(dropna = False)

GENERO
Masculino               3194
Feminino                1056
Prefiro não informar      12
NaN                        9
Name: count, dtype: int64

- Use .value_counts() quando quiser frequência simples.

- Use .groupby(...).nunique() quando quiser contar quantos indivíduos únicos (ou outra coluna única) tem por categoria.

Ou seja, tem uma pessoa marcada como "Feminino" com um ID duplicado — por isso aparece 1056 vezes, mas só 1055 IDs diferentes.

In [72]:
dados[dados['IDADE']>30]['NIVEL'].value_counts()

NIVEL
Sênior    525
Pleno     408
Júnior    274
Name: count, dtype: int64

In [73]:
dados[(dados['IDADE']>30) & (dados['GENERO']== 'Feminino')]['NIVEL'].value_counts()

NIVEL
Pleno     117
Sênior    114
Júnior     96
Name: count, dtype: int64

### Pivot Table = Tabela Dinâmica

In [76]:
pd.pivot_table(dados, values = ['ID'], index = ['GENERO'], columns = ['GESTOR?'], aggfunc='count')

ID     
GESTOR?                0.0  1.0
GENERO                         
Feminino               754  132
Masculino             2212  578
Prefiro não informar     9    2